File: cryptoheir-rs/src/network.rs
//! Network utilities for RPC communication

use crate::{types::TxReceipt, Result};
use alloy::{
    primitives::{Address, Bytes, TxHash, U256},
    providers::{Provider, ProviderBuilder, RootProvider},
    rpc::types::TransactionReceipt,
    transports::http::{Client, Http},
};
use std::collections::HashMap;
use tracing::info;

/// Supported networks with Infura
pub fn get_rpc_url(network: &str, infura_key: Option<&str>) -> Option<String> {
    match network {
        // Ethereum
        "mainnet" | "ethereum" => infura_key.map(|key| format!("https://mainnet.infura.io/v3/{}", key)),
        "sepolia" => infura_key.map(|key| format!("https://sepolia.infura.io/v3/{}", key)),
        "holesky" => infura_key.map(|key| format!("https://holesky.infura.io/v3/{}", key)),

        // Polygon
        "polygon" | "polygon-mainnet" => infura_key.map(|key| format!("https://polygon-mainnet.infura.io/v3/{}", key)),
        "polygon-amoy" => infura_key.map(|key| format!("https://polygon-amoy.infura.io/v3/{}", key)),

        // Arbitrum
        "arbitrum" | "arbitrum-mainnet" => infura_key.map(|key| format!("https://arbitrum-mainnet.infura.io/v3/{}", key)),
        "arbitrum-sepolia" => infura_key.map(|key| format!("https://arbitrum-sepolia.infura.io/v3/{}", key)),

        // Optimism
        "optimism" | "optimism-mainnet" => infura_key.map(|key| format!("https://optimism-mainnet.infura.io/v3/{}", key)),
        "optimism-sepolia" => infura_key.map(|key| format!("https://optimism-sepolia.infura.io/v3/{}", key)),

        // Base
        "base" | "base-mainnet" => infura_key.map(|key| format!("https://base-mainnet.infura.io/v3/{}", key)),
        "base-sepolia" => infura_key.map(|key| format!("https://base-sepolia.infura.io/v3/{}", key)),

        // Linea
        "linea" | "linea-mainnet" => infura_key.map(|key| format!("https://linea-mainnet.infura.io/v3/{}", key)),
        "linea-sepolia" => infura_key.map(|key| format!("https://linea-sepolia.infura.io/v3/{}", key)),

        // Local development
        "localhost" | "anvil" | "hardhat" => Some("http://localhost:8545".to_string()),

        _ => None,
    }
}

/// Create an RPC client
pub async fn create_client(rpc_url: &str) -> Result<RootProvider<Http<Client>>> {
    let provider = ProviderBuilder::new().on_http(rpc_url.parse()?);
    Ok(provider)
}

/// Get chain ID
pub async fn get_chain_id(client: &RootProvider<Http<Client>>) -> Result<u64> {
    Ok(client.get_chain_id().await?)
}

/// Get nonce for an address
pub async fn get_nonce(client: &RootProvider<Http<Client>>, address: Address) -> Result<u64> {
    Ok(client.get_transaction_count(address).await?)
}

/// Get gas prices (returns EIP-1559 or legacy)
pub async fn get_gas_prices(
    client: &RootProvider<Http<Client>>,
) -> Result<(Option<U256>, Option<U256>, Option<U256>)> {
    // Try to get EIP-1559 fee estimates first
    match client.get_fee_history(10, Default::default(), &[]).await {
        Ok(fee_history) => {
            // Get latest base fee
            let base_fee_u128 = fee_history
                .latest_block_base_fee()
                .unwrap_or(1_000_000_000u128); // 1 gwei default
            let base_fee = U256::from(base_fee_u128);

            // Set priority fee (tip)
            let max_priority_fee_per_gas = U256::from(1_500_000_000u64); // 1.5 gwei

            // Max fee = 2x base fee + priority fee (to handle fluctuations)
            let max_fee_per_gas = base_fee * U256::from(2) + max_priority_fee_per_gas;

            info!(
                "Using EIP-1559: max_fee={} gwei, priority_fee={} gwei",
                max_fee_per_gas / U256::from(1_000_000_000u64),
                max_priority_fee_per_gas / U256::from(1_000_000_000u64)
            );

            Ok((
                Some(max_fee_per_gas),
                Some(max_priority_fee_per_gas),
                None,
            ))
        }
        Err(_) => {
            // Fallback to legacy gas price
            let gas_price_u128 = client.get_gas_price().await?;
            let gas_price = U256::from(gas_price_u128);
            info!("Using legacy gas price: {} gwei", gas_price / U256::from(1_000_000_000u64));
            Ok((None, None, Some(gas_price)))
        }
    }
}

/// Estimate gas for a transaction
pub async fn estimate_gas(
    client: &RootProvider<Http<Client>>,
    from: Address,
    to: Option<Address>,
    data: &Bytes,
    value: Option<U256>,
) -> Result<U256> {
    let mut tx = alloy::rpc::types::TransactionRequest::default()
        .from(from)
        .input(data.clone().into());

    if let Some(to_addr) = to {
        tx = tx.to(to_addr);
    }

    if let Some(val) = value {
        tx = tx.value(val);
    }

    let gas_u64 = client.estimate_gas(&tx).await?;
    let gas = U256::from(gas_u64);
    info!("Estimated gas: {}", gas);

    // Add 20% buffer for safety
    let gas_with_buffer = gas * U256::from(120) / U256::from(100);
    info!("Gas with 20% buffer: {}", gas_with_buffer);

    Ok(gas_with_buffer)
}

/// Format Wei to ETH string
pub fn format_eth(wei: U256) -> String {
    let eth = wei.to_string().parse::<f64>().unwrap_or(0.0) / 1e18;
    format!("{:.6}", eth)
}

/// Get a transaction by hash
pub async fn get_transaction(
    client: &RootProvider<Http<Client>>,
    tx_hash: TxHash,
) -> Result<alloy::rpc::types::Transaction> {
    client
        .get_transaction_by_hash(tx_hash)
        .await?
        .ok_or_else(|| eyre::eyre!("Transaction not found"))
}

/// Broadcast a signed transaction
pub async fn broadcast_transaction(
    client: &RootProvider<Http<Client>>,
    signed_tx: &Bytes,
) -> Result<TxHash> {
    let pending_tx = client.send_raw_transaction(signed_tx).await?;
    Ok(*pending_tx.tx_hash())
}

/// Wait for transaction receipt
pub async fn wait_for_receipt(client: &RootProvider<Http<Client>>, tx_hash: TxHash) -> Result<TxReceipt> {
    // Poll for receipt with timeout
    let mut attempts = 0;
    const MAX_ATTEMPTS: u32 = 60; // 5 minutes with 5s intervals
    const POLL_INTERVAL: std::time::Duration = std::time::Duration::from_secs(5);

    info!("Polling for transaction receipt (max {} seconds)...", MAX_ATTEMPTS * 5);

    loop {
        match client.get_transaction_receipt(tx_hash).await? {
            Some(receipt) => {
                info!("Receipt received after {} seconds", attempts * 5);
                return receipt_to_tx_receipt(receipt);
            }
            None => {
                attempts += 1;

                // Log progress every 6 attempts (30 seconds)
                if attempts % 6 == 0 {
                    info!(
                        "Still waiting for receipt... ({}/{} seconds elapsed)",
                        attempts * 5,
                        MAX_ATTEMPTS * 5
                    );
                    println!(
                        "  ⏳ Waiting for block to be mined... ({}/{} seconds)",
                        attempts * 5,
                        MAX_ATTEMPTS * 5
                    );
                }

                if attempts >= MAX_ATTEMPTS {
                    return Err(eyre::eyre!(
                        "Timeout waiting for transaction receipt after {} seconds.\n\
                         \n\
                         The transaction may still be pending. You can:\n\
                         1. Check the transaction on a block explorer\n\
                         2. If using a local node (Anvil), mine a block: cast rpc anvil_mine\n\
                         3. Run the broadcast command again (it will detect the existing transaction)",
                        MAX_ATTEMPTS * 5
                    ));
                }

                tokio::time::sleep(POLL_INTERVAL).await;
            }
        }
    }
}

/// Convert alloy receipt to our TxReceipt type
fn receipt_to_tx_receipt(receipt: TransactionReceipt) -> Result<TxReceipt> {
    Ok(TxReceipt {
        transaction_hash: receipt.transaction_hash,
        block_number: receipt.block_number.ok_or_else(|| eyre::eyre!("No block number"))?,
        block_hash: format!("{:?}", receipt.block_hash.ok_or_else(|| eyre::eyre!("No block hash"))?),
        from: receipt.from,
        to: receipt.to,
        gas_used: receipt.gas_used.to_string(),
        status: if receipt.status() { 1 } else { 0 },
        contract_address: receipt.contract_address,
        metadata: HashMap::new(),
    })
}

File: cryptoheir-go/internal/network/network.go
package network

import (
	"context"
	"fmt"
	"log/slog"
	"math/big"
	"time"

	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/types"
	"github.com/ethereum/go-ethereum"
	"github.com/ethereum/go-ethereum/common"
	coretypes "github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/ethclient"
)

var log *slog.Logger

// SetLogger sets the logger for the network package
func SetLogger(logger *slog.Logger) {
	log = logger
}

// GetRPCURL returns the RPC URL for a given network name
func GetRPCURL(network string, infuraAPIKey string) (string, error) {
	// Check for custom RPC URL first (localhost, custom endpoints)
	if network == "localhost" || network == "anvil" || network == "hardhat" {
		return "http://127.0.0.1:8545", nil
	}

	// If Infura key not provided, can't use predefined networks
	if infuraAPIKey == "" {
		return "", fmt.Errorf("INFURA_API_KEY required for network %s", network)
	}

	// Map network names to Infura endpoints
	networkMap := map[string]string{
		// Ethereum
		"mainnet": fmt.Sprintf("https://mainnet.infura.io/v3/%s", infuraAPIKey),
		"sepolia": fmt.Sprintf("https://sepolia.infura.io/v3/%s", infuraAPIKey),
		"holesky": fmt.Sprintf("https://holesky.infura.io/v3/%s", infuraAPIKey),

		// Polygon
		"polygon-mainnet": fmt.Sprintf("https://polygon-mainnet.infura.io/v3/%s", infuraAPIKey),
		"polygon-amoy":    fmt.Sprintf("https://polygon-amoy.infura.io/v3/%s", infuraAPIKey),

		// Arbitrum
		"arbitrum-mainnet": fmt.Sprintf("https://arbitrum-mainnet.infura.io/v3/%s", infuraAPIKey),
		"arbitrum-sepolia": fmt.Sprintf("https://arbitrum-sepolia.infura.io/v3/%s", infuraAPIKey),

		// Optimism
		"optimism-mainnet": fmt.Sprintf("https://optimism-mainnet.infura.io/v3/%s", infuraAPIKey),
		"optimism-sepolia": fmt.Sprintf("https://optimism-sepolia.infura.io/v3/%s", infuraAPIKey),

		// Base
		"base-mainnet": fmt.Sprintf("https://base-mainnet.infura.io/v3/%s", infuraAPIKey),
		"base-sepolia": fmt.Sprintf("https://base-sepolia.infura.io/v3/%s", infuraAPIKey),

		// Linea
		"linea-mainnet": fmt.Sprintf("https://linea-mainnet.infura.io/v3/%s", infuraAPIKey),
		"linea-sepolia": fmt.Sprintf("https://linea-sepolia.infura.io/v3/%s", infuraAPIKey),
	}

	url, exists := networkMap[network]
	if !exists {
		return "", fmt.Errorf("unknown network: %s", network)
	}

	return url, nil
}

// CreateClient creates an Ethereum RPC client
func CreateClient(ctx context.Context, rpcURL string) (*ethclient.Client, error) {
	client, err := ethclient.DialContext(ctx, rpcURL)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to RPC: %w", err)
	}
	return client, nil
}

// GetChainID returns the chain ID from the RPC endpoint
func GetChainID(ctx context.Context, client *ethclient.Client) (uint64, error) {
	chainID, err := client.ChainID(ctx)
	if err != nil {
		return 0, fmt.Errorf("failed to get chain ID: %w", err)
	}
	return chainID.Uint64(), nil
}

// GetNonce returns the transaction count (nonce) for an address
func GetNonce(ctx context.Context, client *ethclient.Client, address common.Address) (uint64, error) {
	nonce, err := client.PendingNonceAt(ctx, address)
	if err != nil {
		return 0, fmt.Errorf("failed to get nonce: %w", err)
	}
	return nonce, nil
}

// GasPrices holds the gas price information for a transaction
type GasPrices struct {
	MaxFeePerGas         *big.Int // EIP-1559
	MaxPriorityFeePerGas *big.Int // EIP-1559
	GasPrice             *big.Int // Legacy
	IsEIP1559            bool
}

// GetGasPrices fetches gas prices, preferring EIP-1559 if available
func GetGasPrices(ctx context.Context, client *ethclient.Client) (*GasPrices, error) {
	// Try EIP-1559 fee history first
	feeHistory, err := client.FeeHistory(ctx, 1, nil, []float64{50})
	if err == nil && len(feeHistory.BaseFee) > 0 {
		baseFee := feeHistory.BaseFee[len(feeHistory.BaseFee)-1]

		// Priority fee: 1.5 gwei
		priorityFee := new(big.Int).Mul(big.NewInt(15), big.NewInt(1e8)) // 1.5 gwei

		// Max fee: 2 * base_fee + priority_fee
		maxFee := new(big.Int).Mul(baseFee, big.NewInt(2))
		maxFee.Add(maxFee, priorityFee)

		return &GasPrices{
			MaxFeePerGas:         maxFee,
			MaxPriorityFeePerGas: priorityFee,
			IsEIP1559:            true,
		}, nil
	}

	// Fallback to legacy gas price
	gasPrice, err := client.SuggestGasPrice(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get gas price: %w", err)
	}

	return &GasPrices{
		GasPrice:  gasPrice,
		IsEIP1559: false,
	}, nil
}

// EstimateGas estimates gas for a transaction with 20% buffer
func EstimateGas(ctx context.Context, client *ethclient.Client, from common.Address, to *common.Address, data []byte, value *big.Int) (*big.Int, error) {
	msg := ethereum.CallMsg{
		From:  from,
		To:    to,
		Data:  data,
		Value: value,
	}

	gasLimit, err := client.EstimateGas(ctx, msg)
	if err != nil {
		// Try to extract revert reason
		_, callErr := client.CallContract(ctx, msg, nil)
		if callErr != nil {
			// Try to decode the error
			if len(callErr.Error()) > 0 {
				return nil, fmt.Errorf("gas estimation failed: %s", callErr.Error())
			}
		}
		return nil, fmt.Errorf("gas estimation failed: %w", err)
	}

	// Add 20% buffer for safety
	buffer := new(big.Int).Div(big.NewInt(int64(gasLimit)), big.NewInt(5))
	gasWithBuffer := new(big.Int).Add(big.NewInt(int64(gasLimit)), buffer)

	return gasWithBuffer, nil
}

// BroadcastTransaction broadcasts a signed raw transaction
func BroadcastTransaction(ctx context.Context, client *ethclient.Client, signedTx []byte) (common.Hash, error) {
	tx := new(coretypes.Transaction)
	if err := tx.UnmarshalBinary(signedTx); err != nil {
		return common.Hash{}, fmt.Errorf("failed to decode signed transaction: %w", err)
	}

	if err := client.SendTransaction(ctx, tx); err != nil {
		return common.Hash{}, fmt.Errorf("failed to broadcast transaction: %w", err)
	}

	return tx.Hash(), nil
}

// GetTransaction retrieves a transaction by hash
func GetTransaction(ctx context.Context, client *ethclient.Client, txHash common.Hash) (*coretypes.Transaction, bool, error) {
	tx, isPending, err := client.TransactionByHash(ctx, txHash)
	if err != nil {
		return nil, false, err
	}
	return tx, isPending, nil
}

// WaitForReceipt polls for a transaction receipt with timeout
func WaitForReceipt(ctx context.Context, client *ethclient.Client, txHash common.Hash) (*types.TxReceipt, error) {
	timeout := 5 * time.Minute
	interval := 5 * time.Second
	deadline := time.Now().Add(timeout)

	log.Info("Waiting for transaction to be mined", "hash", txHash.Hex())

	lastLog := time.Now()
	for time.Now().Before(deadline) {
		receipt, err := client.TransactionReceipt(ctx, txHash)
		if err == nil {
			// Receipt found
			log.Info("Transaction mined", "block", receipt.BlockNumber.Uint64())

			var contractAddr *common.Address
			if receipt.ContractAddress != (common.Address{}) {
				contractAddr = &receipt.ContractAddress
			}

			txReceipt := &types.TxReceipt{
				TransactionHash: receipt.TxHash,
				BlockNumber:     receipt.BlockNumber.Uint64(),
				BlockHash:       receipt.BlockHash.Hex(),
				From:            common.Address{}, // Not available in receipt
				To:              nil,              // Not available in receipt
				GasUsed:         fmt.Sprintf("%d", receipt.GasUsed),
				Status:          receipt.Status,
				ContractAddress: contractAddr,
				Metadata:        make(map[string]interface{}),
			}

			return txReceipt, nil
		}

		// Log progress every 30 seconds
		if time.Since(lastLog) >= 30*time.Second {
			log.Info("Still waiting for confirmation...")
			lastLog = time.Now()
		}

		// Wait before next poll
		time.Sleep(interval)
	}

	return nil, fmt.Errorf("timeout waiting for transaction receipt after %v", timeout)
}

// FormatEth converts wei to ETH string with 6 decimal places
func FormatEth(wei *big.Int) string {
	if wei == nil {
		return "0"
	}

	// Convert to float for display (wei to ETH = / 1e18)
	weiFloat := new(big.Float).SetInt(wei)
	ethFloat := new(big.Float).Quo(weiFloat, big.NewFloat(1e18))

	return fmt.Sprintf("%.6f ETH", ethFloat)
}
